File: java/src/main/java/com/paradigmas/fp/funciones_puras/FuncionesPuras.java
package com.paradigmas.fp.funciones_puras;

import java.util.List;

/**
 * FUNCIONES PURAS
 * ===============
 * Una función pura:
 *   1. Dado el mismo input, siempre retorna el mismo output.
 *   2. No produce efectos secundarios (no modifica estado externo).
 *
 * Java 26: records, lambdas, method references, var
 */
public class FuncionesPuras {

    // ── FUNCIONES PURAS ──────────────────────────────────────────────
    /** Pura: sólo depende de sus argumentos, sin efectos secundarios. */
    public static int suma(int a, int b) {
        return a + b;
    }

    public static double areaCirculo(double radio) {
        return Math.PI * radio * radio;
    }

    public static String saludar(String nombre) {
        return "Hola, " + nombre + "!";
    }

    /** Pura sobre una lista: retorna nueva lista sin mutar la original. */
    public static List<Integer> duplicar(List<Integer> lista) {
        return lista.stream()
                    .map(x -> x * 2)
                    .toList();   // toList() retorna lista inmutable (Java 16+)
    }

    // ── FUNCIONES IMPURAS ────────────────────────────────────────────
    private static int totalAcumulado = 0;  // estado global mutable

    /** Impura: modifica estado externo. */
    public static int sumaImpura(int a, int b) {
        totalAcumulado += a + b;    // efecto secundario
        return a + b;
    }

    /** Impura: efecto secundario de I/O. */
    public static String saludarImpura(String nombre) {
        System.out.println("Hola, " + nombre + "!");  // efecto secundario
        return "Hola, " + nombre + "!";
    }

    // ── DEMO ─────────────────────────────────────────────────────────
    public static void main(String[] args) {
        System.out.println("=== Funciones Puras ===");
        System.out.println(suma(3, 4));           // siempre 7
        System.out.println(suma(3, 4));           // siempre 7
        System.out.println(areaCirculo(5));
        System.out.println(saludar("Ana"));
        System.out.println(duplicar(List.of(1, 2, 3, 4, 5)));

        System.out.println("\n=== Función Impura ===");
        System.out.println(sumaImpura(3, 4));
        System.out.println("totalAcumulado = " + totalAcumulado);  // cambia!
        System.out.println(sumaImpura(3, 4));
        System.out.println("totalAcumulado = " + totalAcumulado);  // cambia de nuevo!
    }
}

File: python/01_funciones_puras/funciones_puras.py
"""
FUNCIONES PURAS
===============
Una función pura:
  1. Dado el mismo input, siempre retorna el mismo output.
  2. No produce efectos secundarios (no modifica estado externo).
"""

# -----------------------------------------------
# FUNCIÓN PURA: solo depende de sus argumentos
# -----------------------------------------------
def suma(a: int, b: int) -> int:
    return a + b

def area_circulo(radio: float) -> float:
    import math
    return math.pi * radio ** 2

def saludar(nombre: str) -> str:
    return f"Hola, {nombre}!"


# -----------------------------------------------
# FUNCIÓN IMPURA: tiene efectos secundarios
# -----------------------------------------------
total_acumulado = 0  # estado global

def suma_impura(a: int, b: int) -> int:
    global total_acumulado
    total_acumulado += a + b   # modifica estado externo -> IMPURA
    return a + b

def saludar_impura(nombre: str) -> str:
    print(f"Hola, {nombre}!")  # efecto secundario (I/O) -> IMPURA
    return f"Hola, {nombre}!"


# -----------------------------------------------
# DEMO
# -----------------------------------------------
if __name__ == "__main__":
    print("=== Funciones Puras ===")
    print(suma(3, 4))         # siempre 7
    print(suma(3, 4))         # siempre 7
    print(area_circulo(5))
    print(saludar("Ana"))

    print("\n=== Función Impura ===")
    print(suma_impura(3, 4))
    print(f"total_acumulado = {total_acumulado}")  # estado cambia!
    print(suma_impura(3, 4))
    print(f"total_acumulado = {total_acumulado}")  # cambia de nuevo!
